### PR TITLE
CRIO changes to support split filesystem 

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind api-change

#### What this PR does / why we need it:
Implement Image_Fs_Info as part of the separate gc KEP.  We should not merge unless this KEP is approved.  TODO: Add a link.  

Opening this up now to gather feedback on feasibility of this implementation.

Note, since I had a few changes in other areas, I vendored these changes here.  I already opened up a PR into c/storage.  Kubernetes is based on KEP approval.
#### Special notes for your reviewer:
We are having some problems with the integration tests picking up an older version of crictl. I moved the integration tests to https://github.com/cri-o/cri-o/pull/7269 just to move this PR through.
#### Does this PR introduce a user-facing change?

```release-note
Added more file system information in `ImageFsInfo` as part of the garbage collection KEP.
```

### CODE REVIEW

File: server/image_fs_info.go
Comment (sohankunkerkar) [inline server/image_fs_info.go:20]: We can restructure this code a bit in terms of readability. Since you used `GetDiskUsageStats` twice in this function, how about creating a helper function?

```go
func getUsage(path string) (*types.FilesystemUsage, error) {

  bytes, inodes, err := crioStorage.GetDiskUsageStats(path)
  if err != nil {
    return nil, err
  }

  return &types.FilesystemUsage{
    Timestamp: time.Now().UnixNano(),
    FsId: &types.FilesystemIdentifier{Mountpoint: path},
    UsedBytes: &types.UInt64Value{Value: bytes},
    InodesUsed: &types.UInt64Value{Value: inodes},
  }, nil

}
```
```go
func getStorageFsInfo(store storage.Store) (*types.ImageFsInfoResponse, error) {

  rootPath := store.GraphRoot()
  storageDriver := store.GraphDriverName()
  imagePath := store.ImageRoot()
  var graphRootPath string 
  if imagePath == "" {
    graphRootPath = path.Join(rootPath, storageDriver+"-images")  
  } else {
    graphRootPath = path.Join(rootPath, storageDriver+"-containers")
  }

  graphUsage, err := getUsage(graphRoot)
  if err != nil {
    return nil, err
  }

  if imagePath == "" {
    return &types.ImageFsInfoResponse{
      ImageFilesystems: []*types.FilesystemUsage{graphUsage},
      ContainerFilesystems: []*types.FilesystemUsage{graphUsage},  
    }, nil
  }
  resp := &types.ImageFsInfoResponse{
    ContainerFilesystems: []*types.FilesystemUsage{graphUsage},
  }

  if imagePath != "" {

    imageRoot := path.Join(imagePath, storageDriver+"-images")
    
    imageUsage, err := getUsage(imageRoot)
    if err != nil {
      return nil, err
    }

    resp.ImageFilesystems = []*types.FilesystemUsage{imageUsage}

  }

  return resp, nil

}
```

Comment (kannon92) [reply server/image_fs_info.go:20]: Nice suggestion. Will add.

Comment (kannon92) [reply server/image_fs_info.go]: Done.

File: server/image_fs_info.go
Comment (saschagrunert) [inline server/image_fs_info.go:39]: We return above `if imagePath == ""`, so we do not need this additional check.

File: server/image_fs_info.go
Comment (saschagrunert) [inline server/image_fs_info.go:65]: let's wrap the errors:
```suggestion
		return nil, fmt.Errorf("get disk usage for path %s: %w", containerPath, err)
```

**Comment by saschagrunert** [comment]
Do we need an integration test for this change?

**Comment by kannon92** [comment]
> Do we need an integration test for this change?

I'm not really sure how much testing is done around c/storage in this code base. I'll leave that up to you all about whether or not this is needed.

**Comment by kannon92** [comment]
@saschagrunert I followed up with @haircommander.

It turns out that we already have integration tests with imagestore.

The main thing I should as part of this PR is add some integration tests for imagefsinfo.

**Review by saschagrunert — APPROVED** [review]
@kannon92 I'd say we add a test case covering the code paths for the storage info. We can also follow-up on that as you prefer

/hold

Feel free to lift the hold if you'd like to get it merged first.

**Comment by kannon92** [comment]
> @kannon92 I'd say we add a test case covering the code paths for the storage info. We can also follow-up on that as you prefer
> 
> /hold
> 
> Feel free to lift the hold if you'd like to get it merged first.

The one reason I would object is that I will need the latest critools to test this. Since I modified imagefsinfo.

Will integration tests pick up the latest main for crictl?

**Comment by kannon92** [comment]
Updated and added an integration test.

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Thanks!

Unfortunately, the integration tests fail in CI: https://github.com/cri-o/cri-o/actions/runs/7143698636/job/19455753370?pr=7269

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Can you bump cri-tools to v1.29.0 in `scripts/versions`?

**Comment by kannon92** [comment]
Version bumped so we will see now!

**Review by saschagrunert — APPROVED** [review]
/unhold

**Review by saschagrunert — CHANGES_REQUESTED** [review]
Oh, `verify-dependencies` needs to be fixed:

https://github.com/cri-o/cri-o/actions/runs/7182428227/job/19558915346?pr=7269

**Comment by kannon92** [comment]
> Oh, `verify-dependencies` needs to be fixed:
> 
> https://github.com/cri-o/cri-o/actions/runs/7182428227/job/19558915346?pr=7269

I pushed a fix for that.

**Comment by saschagrunert** [comment]
Ah the test failures are legit: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/7269/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1734574155182903296/artifacts/cgroupv2-integration/cri-o-gather/artifacts/testout.txt

**Comment by kannon92** [comment]
> @kannon92 we may have to bump, adding that to dependencies.yaml would be nice as well:
> 
> https://github.com/cri-o/cri-o/blob/68728d0874197b62bf487f70702776df226487df/contrib/test/ci/vars.yml#L17

I updated that also. :(

**Comment by saschagrunert** [comment]
> I updated that also. :(

Thanks! I'm not sure if we have to wait for a new image build or if it applies directly, let's see.

**Comment by saschagrunert** [comment]
@sohankunkerkar do we have to merge it as-is and wait for new images to be rebuilt?

**Comment by saschagrunert** [comment]
/override ci/prow/ci-cgroupv2-integration
/override ci/prow/ci-fedora-integration

/hold

Feel free to lift if we have to merge this one as-is.

**Comment by sohankunkerkar** [comment]
> @sohankunkerkar do we have to merge it as-is and wait for new images to be rebuilt?

Yes, we need to merge the cri-tools specific change first. After that, we can build an image and run CI on this change.

**Comment by kannon92** [comment]
> > @sohankunkerkar do we have to merge it as-is and wait for new images to be rebuilt?
> 
> Yes, we need to merge the cri-tools specific change first. After that, we can build an image and run CI on this change.

Opened https://github.com/cri-o/cri-o/pull/7585 for this.

**Comment by kannon92** [comment]
> > @sohankunkerkar do we have to merge it as-is and wait for new images to be rebuilt?
> 
> Yes, we need to merge the cri-tools specific change first. After that, we can build an image and run CI on this change.

@sohankunkerkar is it possible to do this at some point? We merged the critools but it seems that the CI is still failing.

**Comment by sohankunkerkar** [comment]
> > > @sohankunkerkar do we have to merge it as-is and wait for new images to be rebuilt?
> 
> > 
> 
> > Yes, we need to merge the cri-tools specific change first. After that, we can build an image and run CI on this change.
> 
> 
> 
> @sohankunkerkar is it possible to do this at some point? We merged the critools but it seems that the CI is still failing.

@kannon92 It looks like yesterday's periodic job run already added that change in CI images.

**Comment by kannon92** [comment]
@haircommander @saschagrunert PTAL when you get a chance. @sohankunkerkar and I found a path forward to fix the CI failures so this should be good to go now as is.

**Comment by kannon92** [comment]
@sohankunkerkar I think there is a legit failure here. I'll see if I find the cause.

**Comment by kannon92** [comment]
@haircommander we think that this one needs #7606 as we are hitting caching issues where `crictl` is still picking up the old `1.28` version and we want to force a new clone so we can pick up the latest version.

**Comment by sohankunkerkar** [comment]
hmm... the latest build still uses the older version of `crictl`:
```sh
[skunkerk@ci-op-ygyzh78h-39a11-buildhost ~]$ crictl --version
crictl version 1.28.0

```

**Review by saschagrunert — APPROVED** [review]

**Pull request merged**
